src/GraphQL/Internal/Schema.hs

```diff
@@ -32,21 +32,47 @@ module GraphQL.Internal.Schema
   , AnnotatedType(..)
   , ListType(..)
   , NonNullType(..)
+  , DefinesTypes(..)
+  , doesFragmentTypeApply
   ) where
 
 import Protolude hiding (Type)
 
+import qualified Data.Map as Map
 import GraphQL.Value (Value)
 import GraphQL.Internal.Name (HasName(..), Name, unsafeMakeName)
 
 -- XXX: Use the built-in NonEmptyList in Haskell
-newtype NonEmptyList a = NonEmptyList [a] deriving (Eq, Show)
+newtype NonEmptyList a = NonEmptyList [a] deriving (Eq, Show, Functor, Foldable)
+
+-- | A thing that defines types. Excludes definitions of input types.
+class DefinesTypes t where
```

Comment:
ah right, not sure.  We have a mix, can't really place it anywhere so probably fine

```
$ gg -h "^class " | sort | uniq
class BuildFieldResolver m fieldResolverType where
class Defaultable a where
class FromValue a where
class GenericAnnotatedInputType (f :: Type -> Type) where
class GenericEnumValues (f :: Type -> Type) where
class GenericFromValue (f :: Type -> Type) where
class GenericInputObjectFieldDefinitions (f :: Type -> Type) where
class GraphQLEnum a where
class GraphQLError e where
class HasAnnotatedInputType a where
class HasAnnotatedType a where
class HasArgumentDefinition a where
class HasFieldDefinition a where
class HasFieldDefinitions a where
class HasInterfaceDefinition a where
class HasInterfaceDefinitions a where
class HasName a where
class HasObjectDefinition a where
class HasResolver m a where
class RunFields m a where
class RunUnion m union objects where
class ToValue a where
class UnionTypeObjectTypeDefinitionList a where
```